sled-agent/src/params.rs

```diff
@@ -550,7 +550,27 @@ impl ServiceZoneRequest {
 
     // The name of a unique identifier for the zone, if one is necessary.
     pub fn zone_name_unique_identifier(&self) -> Option<String> {
-        self.dataset.as_ref().map(|d| d.name.pool().to_string())
+        self.dataset.as_ref().map(|d| d.name.pool().to_string()).or_else(|| {
+            match &self.zone_type {
+                // The switch zone is necessarily a singleton.
+                ZoneType::Switch => None,
+
+                // We should never get here for the following zones because they
+                // have a dataset.
+                ZoneType::Clickhouse
+                | ZoneType::CockroachDb
+                | ZoneType::Crucible
+                | ZoneType::ExternalDns
+                | ZoneType::InternalDns => None,
+
+                // Other zones should be able to have multiple instances on the
+                // same Sled.  Not all have been tested yet.
+                ZoneType::Nexus => Some(self.id.to_string()),
+                ZoneType::CruciblePantry
+                | ZoneType::Ntp
+                | ZoneType::Oximeter => None,
+            }
+        })
```

Comment:
I chickened out a little bit here.

- There's a bunch of stuff that assumes `oxz_switch`.  They could be looking for zones of the right type instead, but I didn't want to unpack all that here.
- Similar with `oxz_oximeter`, though that was just in the tests.
- Zones that have a dataset already have their *pool name* in the zone name.  I was reluctant to change this because I don't know what, if anything, depends on it.

Reply:
I can try making this change as a follow-up -- I think consistency and predictability wins here, rather than the convention to match zone names with related resources.

Reply:
Done here: https://github.com/oxidecomputer/omicron/pull/3566